Pass f_locals as a dict rather than kwargs

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98107

Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Pass f_locals as a dict rather than kwargs

Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

voznesenskym commented:
Re-PR of https://github.com/pytorch/pytorch/pull/97703/files as myself.

ezyang commented:
The perf compare run at https://hud.pytorch.org/pytorch/pytorch/pull/97703?sha=e75a7cd2be15ea96f36c31019ae4be460dd5a05d suggests no material change in speedup, or maybe a slight boost.

```
torchbench 2023-03-31T15:22:06.8485112Z speedup             gmean=1.18x mean=1.197x
timm 2 2023-03-31T15:07:06.1972904Z speedup             gmean=1.32x mean=1.353x
timm 1 2023-03-31T14:54:56.7977924Z speedup             gmean=1.38x mean=1.394x
huggingface 2023-03-31T14:52:28.8494698Z speedup             gmean=1.47x mean=1.498x

```

ezyang commented on test/dynamo/test_comptime.py:
```diff
@@ -40,8 +40,9 @@ def _(ctx):
             """\
-def forward(self, x : torch.Tensor):
-    mul = x * 2;  x = None""",
+def forward(self, L_x_ : torch.Tensor):
```
This renaming is truly a travesty. We should figure out how to get the original local as the input name...

ezyang replied:
(Doesn't have to be this PR)

voznesenskym replied:
Lemme play with that, I can stack it

voznesenskym replied:
Stacked.

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -623,11 +622,16 @@ def source_ref(source):
+            local_builder.scope["G"] = global_builder.scope["G"]
         # source_ref can cause a cycle, make sure we break it with weakref
         w_local = weakref.ref(local_builder)
         w_global = weakref.ref(global_builder)
```
Scroll down a bit to line 660 (new) and there's this code

```

        assert not (set(local_builder.argnames) & set(global_builder.argnames))
        # see parallel handling of ".0" / "___implicit0" in _eval_frame.c
        largs = [a for a in local_builder.scope.keys() if a == "___implicit0"]
        largs += [a for a in local_builder.argnames if a != "___implicit0"]
        largs += ["**___kwargs_ignored"]
        args = ",".join(largs)
```

We've removed .0 renaming to __implicit0 so something needs to happen to this code too

voznesenskym replied:
Good eye

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -623,11 +622,16 @@ def source_ref(source):
+        # might get a cross ref between local and global, like L['mod_name'][G['some_key']]
+        # the inverse is illegal.
+        if "G" in global_builder.scope:
+            local_builder.scope["G"] = global_builder.scope["G"]
```
I think you can do this a lot more simply.

Instead of passing in `user_scope` and controlling if it goes to L or G depending on `local`, you should pass in `scope`. Then modify this code:

```
        local_builder = GuardBuilder(
            self.id_ref,
            source_ref,
            combine_scopes(f_globals, f_locals),
            self,
            local=True,
        )
        global_builder = GuardBuilder(
            self.id_ref, source_ref, f_globals, self, local=False
        )
```

to

```
        local_builder = GuardBuilder(
            self.id_ref,
            source_ref,
            {"G": f_globals, "L": f_locals},
            self,
        )
        global_builder = GuardBuilder(
            self.id_ref, source_ref, {"G": f_globals}, self
        )
```

which makes it super clear what's going on: local guards have access to globals and locals, while global guards only have access to globals. (TBH, it's also not altogether clear to me why we need separate local and global guard builder, but whatever...)

voznesenskym replied:
> I think you can do this a lot more simply.

Agreed

> TBH, it's also not altogether clear to me why we need separate local and global guard builder, but whatever...

We have separate builders because of how we do scope binding:

```
exec(py_code, global_builder.scope, out)
```
And
```
guard_fn.global_scope = global_builder.scope
```

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -723,13 +727,12 @@ def compile_check_fn(
-        # print("RUNNING PY CODE", py_code)
         exec(py_code, global_builder.scope, out)
         guard_fn = out["___make_guard_fn"](*closure_vars.values())
         guard_fn.closure_vars = closure_vars
```
Scroll down to line 743 (new). You have assigned scope to global_scope on guard fn. But this is now a `{"G": global_scope}`. It seems like it would make more sense to save scope["G"] instead. Otherwise, there is an implication that there is other stuff in scope you're interested in. But you're only interested in the G key, cf line 780

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -787,11 +791,11 @@ def dict_const_keys(value):
         # To workaround repr(Enum) returning invalid global reference before python 3.11
         # by calling enum_repr and removing quotes to render enum in guard code.
-        const_keys_str = f"{ {enum_repr(k) if isinstance(k, enum.Enum) else repr(k) for k in const_keys} }".replace(
+        const_keys_str = f"{ {enum_repr(k, guard_source) if isinstance(k, enum.Enum) else repr(k) for k in const_keys} }".replace(
```
This still doesn't make any sense. You're feeding in guard_source to make the decision whether or not to look for the enum namespace from globals or locals, but there isn't necessarily any relationship between the guard source. Let's imagine we're doing a key check on `foobar`. This will look like `___dict_const_keys(foobar) == {EnumNs.blah}`. `guard_source` being local or global says something about `foobar`... but you're using that to decide if you will generate `L["EnumNs"]` or `G["EnumNs"]`

Just unconditionally get this from the global namespace and file a TODO to fix it properly...

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -533,7 +535,12 @@ def compile_subgraph(
+            # This will in turn result in spurious variables showing up in the graph.
+            # This was very tricky to debug. For an example, dump the graph at call_user_compiler
+            # while running test_subgraphs.py
+            if isinstance(v.source, LocalSource) and v.source.local_name == k:
```
I think it is a lot clearer if we do `v.source == LocalSource(k)` (this is assuming LocalSource has no subclasses)

voznesenskym replied:
LocalSource has LocalInputSource subclass, but I can clean that up actually, we don't really need it anymore.

voznesenskym replied:
Done

ezyang reviewed (approved):
There are some substantive comments but overall looks good.

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

XuehaiPan commented:
Would it be better to wrap the `f_locals` dictionary with `types.MappingProxyType`? `frame.f_locals` is a `property`. The local variables are not stored in the `f_locals` dict. May explicitly make it read-only to prevent potential bugs.

```python
import sys


def foo():
    a = 1
    frame = sys._getframe()
    print("exec: a = 1")
    print("a                   =", a)
    print('locals()["a"]       =', locals()["a"])
    print('frame.f_locals["a"] =', frame.f_locals["a"])
    print()

    frame.f_locals["a"] = 2
    print('exec: frame.f_locals["a"] = 2')
    print("a                   =", a)
    print('locals()["a"]       =', locals()["a"])
    print('frame.f_locals["a"] =', frame.f_locals["a"])


foo()
# exec: a = 1
# a                   = 1
# locals()["a"]       = 1
# frame.f_locals["a"] = 1
#
# exec: frame.f_locals["a"] = 2
# a                   = 1
# locals()["a"]       = 1
# frame.f_locals["a"] = 1


def bar():
    a = 1
    frame = sys._getframe()
    f_locals = frame.f_locals
    print("exec: a = 1")
    print("f_locals['a']       =", f_locals["a"])
    print("frame.f_locals['a'] =", frame.f_locals["a"])
    print()

    a = 2
    print("exec: a = 2")
    print("f_locals['a']       =", f_locals["a"])
    print("frame.f_locals['a'] =", frame.f_locals["a"])
    print()

    assert f_locals is frame.f_locals
    print("exec: assert f_locals is frame.f_locals")
    print("f_locals['a']       =", f_locals["a"])
    print("frame.f_locals['a'] =", frame.f_locals["a"])
    print()


bar()
# exec: a = 1
# f_locals['a']       = 1
# frame.f_locals['a'] = 1
#
# exec: a = 2
# f_locals['a']       = 1
# frame.f_locals['a'] = 2
#
# exec: assert f_locals is frame.f_locals
# f_locals['a']       = 2
# frame.f_locals['a'] = 2
```

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
DevExp friends! If a Meta internal breakage takes you here, do not revert. Instead, disable the internal CI. It is almost certainly relying on a string representation of output which IS NOT A VALID CONTRACT offered by PT2. Instead, disable the test and instruct the test owner to rewrite the test against the BC guard object in torch/_guards.py.

ezyang commented:
@XuehaiPan the safety would cost us another allocation in the guard path. IDK, doesn't seem like a good tradeoff, since the guard generation code is all TCB

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge -f "Passing, flaky test"
